migrator.py

```diff
@@ -3,39 +3,153 @@
+                contact_id_2_nameset[contact_id]=set([person_name])
+            # Note: the difference between learn_contact_ids and this version is we don't add the name to the set
+    except KeyError:
+        # likely no Contacts2 section
+        pass
+    except Exception as e:
+        logging.warning("Exception while learning contact ids")
+        logging.exception(str(e))
+
+def migrate_directories_under(input_root_dir: Path, 
+                              db: DigikamDb, 
+                              dry_run: bool,
+                              contacts_file: Path):
     """Traverses directory tree to find directories to migrate."""
+
+    # Build a contact_id to name(s) dictionary
+    global_names: GlobalNames = {}
+
+    if contacts_file is not None:
+        # contacts.xml seems more authoratative than *.ini files
+        # Use this whenever possible.
+        # Can be found at %LocalAppData%\Google\Picasa2\contacts\contacts.xml
+        # Also can be obtained from performing a backup operation from 
+        # within Picasa.  It will be in the backup location as 
+        # $Application Data\Google\Picasa2\contacts\backup.xml
+        logging.info(f'Reading contacts from {contacts_file}')
+        tree = ET.parse(contacts_file)
+        for contact in tree.getroot():
+            logging.info(f'{contact.attrib}')
+            global_names[contact.attrib['id']] = contact.attrib['name']
+
+    else:
+        # Alternate way using *.ini files
+        # Sometimes names will be missing or conflicting if found 
+        # different in multiple directories.  Conflicting names
+        # will be merged by this script by concatenating them with " | " 
+        # separators in sorted order.  Missing names will have an
+        # auto generated value of  
+        contact_id_2_nameset: Dict[str, set] = {}
+        for input_dir, subdirs, files in os.walk(input_root_dir):
+            for ini_file in (_PICASA_INI_FILE,_OLD_PICASA_INI_FILE):
+                learn_contact_ids(input_dir,ini_file,contact_id_2_nameset)
+        for input_dir, subdirs, files in os.walk(input_root_dir):
+            for ini_file in (_PICASA_INI_FILE,_OLD_PICASA_INI_FILE):
+                learn_old_contact_ids(input_dir,ini_file,contact_id_2_nameset)
+
+        logging.info("contact_id_2_nameset")
+        logging.info("------------------------------------------------------------")
+        try:
+            for contact_id, names in contact_id_2_nameset.items():
+                person_name = ""
+                for i, item in enumerate(sorted(names)):
+                    person_name += item
+                    if i < (len(names)-1):
+                        # ambiguous contact_id will have a concatenated list of names
+                        person_name += "|"
+                logging.info(f"'{contact_id}': '{person_name}'")
+                # Create or add contacts
+                global_names[contact_id] = person_name
+        except Exception as e:
```

Comment:
Why catch here?

Reply:
Had an unexpected exception before this version of code.  Probably OK to remove the handler now.